File: src/main/java/com/ra4king/circuitsim/gui/ComponentPeer.java
package com.ra4king.circuitsim.gui;

import java.util.List;

import com.ra4king.circuitsim.gui.Connection.PortConnection;
import com.ra4king.circuitsim.simulator.CircuitState;
import com.ra4king.circuitsim.simulator.Component;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.paint.Color;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * @author Roi Atalla
 */
public abstract class ComponentPeer<C extends Component> extends GuiElement {
	private C component;
	private Properties properties;
	private List<PortConnection> connections;
	
	public ComponentPeer(int x, int y, int width, int height) {
		super(x, y, width, height);
	}
	
	protected final void init(C component, Properties properties, List<PortConnection> connections) {
		if (this.component != null) {
			throw new IllegalStateException("ComponentPeer already initialized.");
		}
		
		this.component = component;
		this.properties = properties;
		this.connections = connections;
	}
	
	public @NotNull C getComponent() {
		return component;
	}
	
	public Properties getProperties() {
		return properties;
	}
	
	@NotNull
	@Override
	public List<PortConnection> getConnections() {
		return connections;
	}

	@Override
	public void paint(GraphicsContext graphics, @Nullable CircuitState circuitState) {
		GuiUtils.drawName(graphics, this, getProperties().getValue(Properties.LABEL_LOCATION));
		graphics.setFill(Color.WHITE);
		graphics.setStroke(Color.BLACK);
		GuiUtils.drawShape(graphics::fillRect, this);
		GuiUtils.drawShape(graphics::strokeRect, this);
	}

	@Override
	public String toString() {
		return getComponent().toString();
	}
}

File: src/main/kotlin/com/ra4king/circuitsim/gui/ComponentPeer.kt
package com.ra4king.circuitsim.gui

import com.ra4king.circuitsim.gui.Connection.PortConnection
import com.ra4king.circuitsim.gui.GuiUtils.drawName
import com.ra4king.circuitsim.gui.GuiUtils.drawShape
import com.ra4king.circuitsim.simulator.CircuitState
import com.ra4king.circuitsim.simulator.Component
import javafx.scene.canvas.GraphicsContext
import javafx.scene.paint.Color

/**
 * @author Roi Atalla
 */
abstract class ComponentPeer<C : Component>(x: Int, y: Int, width: Int, height: Int) : GuiElement(x, y, width, height) {

    lateinit var component: C
    lateinit var properties: Properties
        private set
    override lateinit var connections: MutableList<PortConnection>

    protected fun init(component: C, properties: Properties, connections: MutableList<PortConnection>) {
        this.component = component
        this.properties = properties
        this.connections = connections
    }

    override fun paint(graphics: GraphicsContext, circuitState: CircuitState?) {
        drawName(graphics, this, this.properties.getValue(Properties.LABEL_LOCATION))
        graphics.fill = Color.WHITE
        graphics.stroke = Color.BLACK
        drawShape(this) { v: Double, v1: Double, v2: Double, v3: Double -> graphics.fillRect(v, v1, v2, v3) }
        drawShape(this) { v: Double, v1: Double, v2: Double, v3: Double -> graphics.strokeRect(v, v1, v2, v3) }
    }

    override fun toString() = component.toString()
}
